src/metabase/api/user.clj

```diff
@@ -248,36 +262,48 @@
     (not ldap_auth))))
 
 (api/defendpoint PUT "/:id"
-  "Update an existing, active `User`."
-  [id :as {{:keys [email first_name last_name group_ids is_superuser login_attributes locale] :as body} :body}]
-  {email            (s/maybe su/Email)
-   first_name       (s/maybe su/NonBlankString)
-   last_name        (s/maybe su/NonBlankString)
-   group_ids        (s/maybe [su/IntGreaterThanZero])
-   is_superuser     (s/maybe s/Bool)
-   login_attributes (s/maybe user/LoginAttributes)
-   locale           (s/maybe su/ValidLocale)}
-  (check-self-or-superuser id)
+  "Update an existing, active `User`.
+  Self or superusers can update user info and groups.
+  Group Managers can only add/remove users from groups they are manager of."
+  [id :as {{:keys [email first_name last_name user_group_memberships
+                   is_superuser is_group_manager login_attributes locale] :as body} :body}]
+  {email                  (s/maybe su/Email)
+   first_name             (s/maybe su/NonBlankString)
+   last_name              (s/maybe su/NonBlankString)
+   user_group_memberships (s/maybe [user/UserGroupMembership])
+   is_superuser           (s/maybe s/Bool)
+   is_group_manager       (s/maybe s/Bool)
+   login_attributes       (s/maybe user/LoginAttributes)
+   locale                 (s/maybe su/ValidLocale)}
+  (try
+   (check-self-or-superuser id)
+   (catch clojure.lang.ExceptionInfo _e
+     (validation/check-group-manager)))
+
   ;; only allow updates if the specified account is active
   (api/let-404 [user-before-update (fetch-user :id id, :is_active true)]
     ;; Google/LDAP non-admin users can't change their email to prevent account hijacking
     (api/check-403 (valid-email-update? user-before-update email))
     ;; can't change email if it's already taken BY ANOTHER ACCOUNT
     (api/checkp (not (db/exists? User, :%lower.email (if email (u/lower-case-en email) email), :id [:not= id]))
-      "email" (tru "Email address already associated to another user."))
+                "email" (tru "Email address already associated to another user."))
     (db/transaction
-      (api/check-500
-       (db/update! User id
-         (u/select-keys-when body
-           :present (into #{:locale} (when api/*is-superuser?* [:login_attributes]))
-           :non-nil (set (concat [:first_name :last_name :email]
-                                 (when api/*is-superuser?*
-                                   [:is_superuser]))))))
-      (maybe-set-user-permissions-groups! id group_ids is_superuser)
-      (maybe-update-user-personal-collection-name! user-before-update first_name last_name)))
+     ;; only user or self can update user info
+     ;; implicitly prevent group manager from updating users' info
+     (when (or (= id api/*current-user-id*)
```

Comment:
It's not a good idea to silently _not_ update user info if called by Group Manager and still returns 200.
But it seems like a rare use-case so I'm living it this way.